File: apps/api/app/schemas/episode.py
from datetime import datetime
from uuid import UUID
from typing import Optional
from pydantic import BaseModel, ConfigDict, Field

from app.schemas.common import PaginationParams, PaginatedResponse


class EpisodeConstraints(BaseModel):
    max_duration_minutes: int = Field(default=30, ge=1, le=120)
    allowed_techniques: list[str] = Field(default_factory=list)
    safety_level: str = Field(default="active", pattern="^(passive|active|aggressive)$")
    max_concurrent_attacks: int = Field(default=3, ge=1, le=10)
    stop_on_first_detection: bool = False
    stop_on_successful_attack: bool = False


class EpisodeBase(BaseModel):
    scenario: str
    constraints: EpisodeConstraints = Field(default_factory=EpisodeConstraints)


class EpisodeCreate(EpisodeBase):
    project_id: UUID
    target_app_id: UUID


class EpisodeUpdate(BaseModel):
    status: Optional[str] = Field(default=None, pattern="^(pending|initializing|running|completed|failed|cancelled)$")
    error: Optional[str] = None


class EpisodeResponse(EpisodeBase):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    project_id: UUID
    target_app_id: UUID
    status: str
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    error: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class EpisodeDetail(EpisodeResponse):
    target_app: Optional[dict] = None
    attacks: list[dict] = Field(default_factory=list)
    detections: list[dict] = Field(default_factory=list)
    responses: list[dict] = Field(default_factory=list)
    score: Optional[dict] = None


class EpisodeListParams(PaginationParams):
    project_id: Optional[UUID] = None
    target_app_id: Optional[UUID] = None
    scenario: Optional[str] = None
    status: Optional[str] = Field(default=None, pattern="^(pending|initializing|running|completed|failed|cancelled)$")
    date_from: Optional[datetime] = None
    date_to: Optional[datetime] = None


class EpisodeListResponse(PaginatedResponse[EpisodeResponse]):
    pass

File: packages/shared/src/schemas/episode.ts
import { z } from "zod";
import { UUIDSchema, DateTimeSchema, PaginationParamsSchema, PaginatedResponseSchema } from "./common";

export const EpisodeStatusSchema = z.enum([
  "pending",
  "initializing",
  "running",
  "completed",
  "failed",
  "cancelled"
]);
export type EpisodeStatus = z.infer<typeof EpisodeStatusSchema>;

export const SafetyLevelSchema = z.enum(["passive", "active", "aggressive"]);
export type SafetyLevel = z.infer<typeof SafetyLevelSchema>;

export const EpisodeConstraintsSchema = z.object({
  maxDurationMinutes: z.number().int().positive().max(120).default(30),
  allowedTechniques: z.array(z.string()).default([]),
  safetyLevel: SafetyLevelSchema.default("active"),
  maxConcurrentAttacks: z.number().int().positive().default(3),
  stopOnFirstDetection: z.boolean().default(false),
  stopOnSuccessfulAttack: z.boolean().default(false)
});
export type EpisodeConstraints = z.infer<typeof EpisodeConstraintsSchema>;

export const EpisodeSchema = z.object({
  id: UUIDSchema,
  projectId: UUIDSchema,
  targetAppId: UUIDSchema,
  scenario: z.string(),
  status: EpisodeStatusSchema.default("pending"),
  constraints: EpisodeConstraintsSchema,
  startedAt: DateTimeSchema.optional(),
  completedAt: DateTimeSchema.optional(),
  error: z.string().optional(),
  createdAt: DateTimeSchema,
  updatedAt: DateTimeSchema
});
export type Episode = z.infer<typeof EpisodeSchema>;

export const CreateEpisodeRequestSchema = z.object({
  projectId: UUIDSchema,
  targetAppId: UUIDSchema,
  scenario: z.string(),
  constraints: EpisodeConstraintsSchema.partial().optional()
});
export type CreateEpisodeRequest = z.infer<typeof CreateEpisodeRequestSchema>;

export const EpisodeListParamsSchema = PaginationParamsSchema.extend({
  projectId: UUIDSchema.optional(),
  targetAppId: UUIDSchema.optional(),
  scenario: z.string().optional(),
  status: EpisodeStatusSchema.optional(),
  dateFrom: DateTimeSchema.optional(),
  dateTo: DateTimeSchema.optional()
});
export type EpisodeListParams = z.infer<typeof EpisodeListParamsSchema>;

export const EpisodeListResponseSchema = PaginatedResponseSchema(EpisodeSchema);
export type EpisodeListResponse = z.infer<typeof EpisodeListResponseSchema>;

export const EpisodeDetailSchema = EpisodeSchema.extend({
  targetApp: z.object({
    id: UUIDSchema,
    name: z.string(),
    type: z.string()
  }).optional(),
  attacks: z.array(z.object({
    id: UUIDSchema,
    techniqueId: z.string(),
    owaspCategory: z.string(),
    success: z.boolean(),
    confidence: z.number().min(0).max(1),
    timestamp: DateTimeSchema
  })).default([]),
  detections: z.array(z.object({
    id: UUIDSchema,
    attackId: UUIDSchema.optional(),
    detected: z.boolean(),
    detectionType: z.string(),
    confidence: z.number().min(0).max(1),
    timestamp: DateTimeSchema
  })).default([]),
  responses: z.array(z.object({
    id: UUIDSchema,
    detectionId: UUIDSchema.optional(),
    actionType: z.string(),
    success: z.boolean(),
    timestamp: DateTimeSchema
  })).default([]),
  score: z.object({
    detectionRate: z.number().min(0).max(1),
    mttrSeconds: z.number().nonnegative(),
    coverage: z.record(z.number().min(0).max(1)),
    overallScore: z.number().min(0).max(100)
  }).optional()
});
export type EpisodeDetail = z.infer<typeof EpisodeDetailSchema>;